pkg/vm/builtins/modulo.go

```diff
@@ -0,0 +1,542 @@
+		if err != nil {
+			return err
+		}
+
+		offset, err := mem.ReadAsAddress(&addr)
+		if err != nil {
+			return err
+		}
+
+		for copyI := 0; copyI < int(nCopies); copyI++ {
+			copyAddr, err := offsetsPtr.AddOffset(int16(3*(index+uint64(copyI)) + uint64(i)))
+			if err != nil {
+				return err
+			}
+			mv := memory.MemoryValueFromMemoryAddress(&offset)
+			if err := mem.WriteToAddress(&copyAddr, &mv); err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+// Given a value, writes its n_words to memory, starting at address = addr.
+func (m *ModBuiltin) writeNWordsValue(mem *memory.Memory, addr memory.MemoryAddress, value big.Int) error {
+	for i := 0; i < N_WORDS; i++ {
+		word := new(big.Int).Mod(&value, &m.shift)
+		modAddr, err := addr.AddOffset(int16(i))
+		if err != nil {
+			return err
+		}
+		mv := memory.MemoryValueFromFieldElement(new(fp.Element).SetBigInt(word))
+		if err := mem.WriteToAddress(&modAddr, &mv); err != nil {
+			return err
+		}
+		value.Div(&value, &m.shift)
+	}
+	if value.Sign() != 0 {
+		return fmt.Errorf("writeNWordsValue: value should be zero")
+	}
+	return nil
+}
+
+// Fills a value in the values table, if exactly one value is missing.
+// Returns true on success or if all values are already known.
+// Given known, res, p fillValue tries to compute the minimal integer operand x which
+// satisfies the equation op(x,known) = res + k*p for some k in {0,1,...,self.k_bound-1}.
+func (m *ModBuiltin) fillValue(mem *memory.Memory, inputs ModBuiltinInputs, index int, op ModBuiltinType) (bool, error) {
+	addresses := make([]memory.MemoryAddress, 0, 3)
+	values := make([]*big.Int, 0, 3)
+
+	for i := 0; i < 3; i++ {
+		addr, err := inputs.offsetsPtr.AddOffset(int16(3*index + i))
+		if err != nil {
+			return false, err
+		}
+		offsetFelt, err := mem.ReadAsElement(addr.SegmentIndex, addr.Offset)
+		if err != nil {
+			return false, err
+		}
+		offset := offsetFelt.Uint64()
+		addr, err = inputs.valuesPtr.AddOffset(int16(offset))
+		if err != nil {
+			return false, err
+		}
+		addresses = append(addresses, addr)
+		// do not check for error, as the value might not be in memory
+		_, value, _ := m.readNWordsValue(mem, addr)
+		values = append(values, value)
+	}
+
+	a, b, c := values[0], values[1], values[2]
+
+	// 2 ** 384 (max value that can be stored in 4 felts)
+	intLim := new(big.Int).Lsh(big.NewInt(1), uint(m.wordBitLen)*N_WORDS)
+	kBound := m.kBound
+	if kBound == nil {
+		kBound = new(big.Int).Set(intLim)
+	}
+
+	switch {
+	case a != nil && b != nil && c == nil:
+		var value big.Int
+		if op == Add {
+			value = *new(big.Int).Add(a, b)
+		} else {
+			value = *new(big.Int).Mul(a, b)
+		}
+		if new(big.Int).Sub(&value, new(big.Int).Mul((new(big.Int).Sub(kBound, big.NewInt(1))), &inputs.p)).Cmp(new(big.Int).Sub(intLim, big.NewInt(1))) == 1 {
+			return false, fmt.Errorf("%s builtin: Expected a %s b - %d * p <= %d", m.String(), m.modBuiltinType, kBound.Sub(kBound, big.NewInt(1)), intLim.Sub(intLim, big.NewInt(1)))
```

Comment:
omg this is so difficult to read 😅 . Is it possible to separate the values? I see you use some of them in the error message, perhaps have it as separate variables so the code is easier to understand

Reply:
But those variables won't really be used anywhere else. I suppose I can write a comment on top of it that represents the mathematical form of this expression.

Reply:
Okay. Merging this now.